### PR TITLE
Block collections of scalars with converters on Cosmos

### LINKED ISSUE #34218: Cosmos: block collections over value-converted elements
We're lacking various things to enable this scenario; the following two have been identified but there may be others:

* #34026 
* #34153

We should detect this scenario during validation, and throw in 9.

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #34218

### CODE REVIEW

File: src/EFCore.Cosmos/Infrastructure/Internal/CosmosModelValidator.cs [hunk 1]
```diff
@@ -41,6 +41,50 @@ public override void Validate(IModel model, IDiagnosticsLogger<DbLoggerCategory.
...
+        IDiagnosticsLogger<DbLoggerCategory.Model.Validation> logger)
+    {
+        foreach (var entityType in model.GetEntityTypes())
+        {
+            ValidateType(entityType, logger);
+        }
+
+        static void ValidateType(ITypeBase typeBase, IDiagnosticsLogger<DbLoggerCategory.Model.Validation> logger)
+        {
+            foreach (var property in typeBase.GetDeclaredProperties())
```
Comment (roji) [inline src/EFCore.Cosmos/Infrastructure/Internal/CosmosModelValidator.cs:64]: Unrelated: could be good to have a utility extension for getting all declared properties recursively (maybe it already even exists - I feel like I've done this several times already, worth checking).

File: src/EFCore.Cosmos/Infrastructure/Internal/CosmosModelValidator.cs [hunk 2]
```diff
@@ -41,6 +41,50 @@ public override void Validate(IModel model, IDiagnosticsLogger<DbLoggerCategory.
...
+            {
+                var typeMapping = property.GetElementType()?.GetTypeMapping();
```
Comment (roji) [inline src/EFCore.Cosmos/Infrastructure/Internal/CosmosModelValidator.cs:66]: Not sure, but maybe it's also worth blocking if a vlue converter is defined on the collection property itself - not sure that works either...

File: test/EFCore.Cosmos.FunctionalTests/Query/NonSharedPrimitiveCollectionsQueryCosmosTest.cs [hunk 3]
```diff
@@ -270,51 +272,30 @@ public override async Task Array_of_bool()
 
     public override async Task Array_of_Guid()
     {
-        await base.Array_of_Guid();
+        Assert.Equal(
```
Comment (roji) [inline test/EFCore.Cosmos.FunctionalTests/Query/NonSharedPrimitiveCollectionsQueryCosmosTest.cs:275]: I think I wasn't aware we were handling e.g. GUID arrays via value converters... Do you think we're maybe blocking too much here? In other words, if it's only queries over such collections that don't work, maybe we shouldn't block the whole thing (at model validation) to allow at least serialization/deserialization? Does the update pipeline handle these correctly (i.e. #34153), and what actually worked in 8.0?

**Review by roji — APPROVED** [review]
The change looks good, but see my comment about being unsure whether we're blocking too much (e.g. GUID arrays).. We may want to have a quick design chat here if you want.

File: test/EFCore.Cosmos.FunctionalTests/Query/NonSharedPrimitiveCollectionsQueryCosmosTest.cs (hunk 3 above)
Comment (ajcvickers) [reply test/EFCore.Cosmos.FunctionalTests/Query/NonSharedPrimitiveCollectionsQueryCosmosTest.cs:275]: Arrays of GUIDs were not supported in 8--throws in model validation--so I think we're okay to continue not supporting them in 9. That being said, if you think the query part for GUIDs converted to strings is something that works/can be made to work in 9, then we can specifically allow just that. Same for enums.

**Pull request merged**